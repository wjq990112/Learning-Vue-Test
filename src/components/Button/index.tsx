import Button from './Button';

export default Button;
export { Button };
export type { ButtonType, ButtonSize } from './Button';
