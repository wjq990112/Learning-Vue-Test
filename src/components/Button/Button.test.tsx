import { shallowMount } from '@vue/test-utils';

import Button from './Button';

describe('Button', () => {
  it('should emit click event.', async () => {
    const wrapper = shallowMount(Button);

    await wrapper.trigger('click');
    expect(wrapper.emitted('click')).toBeDefined();
  });

  it('should not emit click event when disabled.', async () => {
    const wrapper = shallowMount(Button, {
      props: {
        disabled: true,
      },
    });

    await wrapper.trigger('click');
    expect(wrapper.emitted('click')).toBeUndefined();
  });

  it('should not emit click event when loading.', async () => {
    const wrapper = shallowMount(Button, {
      props: {
        loading: true,
      },
    });

    await wrapper.trigger('click');
    expect(wrapper.emitted('click')).toBeUndefined();
  });

  it('should render icon correctly.', async () => {
    const wrapper = shallowMount(Button, {
      props: {
        icon: 'icon-rocket',
      },
    });

    expect(wrapper.find('.icon-sync').exists()).toBeFalsy();
    expect(wrapper.find('.icon-rocket').exists()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();

    await wrapper.setProps({ loading: true });

    expect(wrapper.find('.icon-sync').exists()).toBeTruthy();
    expect(wrapper.find('.icon-rocket').exists()).toBeFalsy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render text correctly.', async () => {
    const wrapper = shallowMount(Button, {
      props: {
        loadingText: 'Custom Loading',
      },
      slots: {
        default: () => <span>Custom Text</span>,
      },
    });

    expect(wrapper.find('span').exists()).toBeTruthy();
    expect(wrapper.find('span').element).toHaveTextContent(/custom text/gi);
    expect(wrapper.find('.icon-sync').exists()).toBeFalsy();
    expect(wrapper.html()).toMatchSnapshot();

    await wrapper.setProps({ loading: true });

    expect(wrapper.find('span').exists()).toBeTruthy();
    expect(wrapper.find('span').element).toHaveTextContent(/custom loading/gi);
    expect(wrapper.find('.icon-sync').exists()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render icon slot correctly.', async () => {
    const wrapper = shallowMount(Button, {
      slots: {
        icon: () => <i class="iconfont icon-rocket" />,
      },
    });

    expect(wrapper.find('.icon-sync').exists()).toBeFalsy();
    expect(wrapper.find('.icon-rocket').exists()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();

    await wrapper.setProps({ loading: true });

    expect(wrapper.find('.icon-sync').exists()).toBeTruthy();
    expect(wrapper.find('.icon-rocket').exists()).toBeFalsy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render loading slot correctly.', async () => {
    const wrapper = shallowMount(Button, {
      slots: {
        loadingIcon: () => <i class="iconfont icon-reload" />,
        loadingText: () => <span>Custom Loading</span>,
      },
    });

    expect(wrapper.find('.icon-reload').exists()).toBeFalsy();
    expect(wrapper.find('span').exists()).toBeFalsy();
    expect(wrapper.html()).toMatchSnapshot();

    await wrapper.setProps({ loading: true });

    expect(wrapper.find('.icon-reload').exists()).toBeTruthy();
    expect(wrapper.find('span').element).toHaveTextContent(/custom loading/gi);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render small button correctly.', () => {
    const wrapper = shallowMount(Button, {
      props: {
        size: 'sm',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render large button correctly.', () => {
    const wrapper = shallowMount(Button, {
      props: {
        size: 'lg',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render primary button correctly.', () => {
    const wrapper = shallowMount(Button, {
      props: {
        type: 'primary',
      },
    });

    expect(wrapper.element).toHaveClass('bg-blue-500');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render success button correctly.', () => {
    const wrapper = shallowMount(Button, {
      props: {
        type: 'success',
      },
    });

    expect(wrapper.element).toHaveClass('bg-green-500');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render warning button correctly.', () => {
    const wrapper = shallowMount(Button, {
      props: {
        type: 'warning',
      },
    });

    expect(wrapper.element).toHaveClass('bg-yellow-500');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render danger button correctly.', () => {
    const wrapper = shallowMount(Button, {
      props: {
        type: 'danger',
      },
    });

    expect(wrapper.element).toHaveClass('bg-red-500');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render block button correctly.', () => {
    const wrapper = shallowMount(Button, {
      props: {
        block: true,
      },
    });

    expect(wrapper.element).toHaveClass('block');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render round button correctly.', () => {
    const wrapper = shallowMount(Button, {
      props: {
        round: true,
      },
    });

    expect(wrapper.element).toHaveClass('rounded-full');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
