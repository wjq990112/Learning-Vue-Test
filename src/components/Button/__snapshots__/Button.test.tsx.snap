// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render block button correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider block w-full text-dark bg-white border-gray-300 active:bg-gray-200\\">
  <!---->
  <!---->
</button>"
`;

exports[`Button should render danger button correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-light bg-red-500 border-transparent active:bg-red-600\\">
  <!---->
  <!---->
</button>"
`;

exports[`Button should render icon correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\"><i class=\\"iconfont inline-block icon-rocket\\"></i>
  <!---->
</button>"
`;

exports[`Button should render icon correctly. 2`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\"><i class=\\"iconfont inline-block animate-spin icon-sync\\"></i>
  <!---->
</button>"
`;

exports[`Button should render icon slot correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\"><i class=\\"iconfont icon-rocket\\"></i>
  <!---->
</button>"
`;

exports[`Button should render icon slot correctly. 2`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\"><i class=\\"iconfont inline-block animate-spin icon-sync\\"></i>
  <!---->
</button>"
`;

exports[`Button should render large button correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-20 min-h-13 px-4 text-lg tracking-widest text-dark bg-white border-gray-300 active:bg-gray-200\\">
  <!---->
  <!---->
</button>"
`;

exports[`Button should render loading slot correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\">
  <!---->
  <!---->
</button>"
`;

exports[`Button should render loading slot correctly. 2`] = `"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\"><i class=\\"iconfont icon-reload\\"></i><span>Custom Loading</span></button>"`;

exports[`Button should render primary button correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-light bg-blue-500 border-transparent active:bg-blue-600\\">
  <!---->
  <!---->
</button>"
`;

exports[`Button should render round button correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider rounded-full text-dark bg-white border-gray-300 active:bg-gray-200\\">
  <!---->
  <!---->
</button>"
`;

exports[`Button should render small button correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-12 min-h-7 px-1 text-sm tracking-wide text-dark bg-white border-gray-300 active:bg-gray-200\\">
  <!---->
  <!---->
</button>"
`;

exports[`Button should render success button correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-light bg-green-500 border-transparent active:bg-green-600\\">
  <!---->
  <!---->
</button>"
`;

exports[`Button should render text correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\">
  <!----><span>Custom Text</span>
</button>"
`;

exports[`Button should render text correctly. 2`] = `"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\"><i class=\\"iconfont inline-block animate-spin icon-sync\\"></i><span class=\\"mx-0.5\\">Custom Loading</span></button>"`;

exports[`Button should render warning button correctly. 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-light bg-yellow-500 border-transparent active:bg-yellow-600\\">
  <!---->
  <!---->
</button>"
`;
